vllm/entrypoints/openai/serving_chat.py

```diff
@@ -364,13 +436,49 @@ async def chat_completion_stream_generator(
                         data = chunk.model_dump_json(exclude_unset=True)
                         yield f"data: {data}\n\n"
                     else:
+                        # check to make sure we haven't "forgotten" to stream
+                        #   any tokens that were generated but previously
+                        #   matched by partial json parsing
+                        # only happens if we are NOT using guided decoding
+                        if tool_parser:
+                            index = len(
+                                tool_parser.prev_tool_call_arr) - 1 if len(
+                                    tool_parser.prev_tool_call_arr) > 0 else 0
+                        else:
+                            index = 0
+                        if (delta_message.tool_calls
+                                and delta_message.tool_calls[0]
+                                and delta_message.tool_calls[0].function and
+                            (delta_message.tool_calls[0].function.arguments
+                             == '' or
+                             delta_message.tool_calls[0].function.arguments and
+                             (output.finish_reason == 'stop'
+                              or output.finish_reason == 'tool_calls'))
+                                and tool_parser
+                                and request.tool_choice == 'auto'):
```

Comment:
Please break apart these conditions into helper functions or variables for subclauses, this is untenable

Reply:
on it :)